### PR TITLE
test: add more tests for popcount

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Review by chfast — COMMENTED** [review]
Cases with single bit are missing.
Same cases should be added to `TEST(execute_numeric, i{32,64}_popcnt)`.

**Comment by axic** [comment]
> Cases with single bit are missing.

Added two single bit cases.

> Same cases should be added to TEST(execute_numeric, i{32,64}_popcnt).

At this point though the tests should be changed to a map then.

File: test/unittests/cxx20_bit_test.cpp [hunk 1]
```diff
@@ -47,20 +47,33 @@ TEST(cxx20_bit, bit_cast_uint32_to_array)
 
 TEST(cxx20_bit, popcount32)
 {
-    EXPECT_EQ(popcount(uint32_t{0}), 0);
-    EXPECT_EQ(popcount(uint32_t{0xffffffff}), 32);
-    EXPECT_EQ(popcount(uint32_t{0xffff0000}), 16);
-    EXPECT_EQ(popcount(uint32_t{0x0000ffff}), 16);
-    EXPECT_EQ(popcount(uint32_t{0x00ffff00}), 16);
-    EXPECT_EQ(popcount(uint32_t{0x00ff00ff}), 16);
+    constexpr std::pair<uint32_t, uint32_t> test_cases[]{
```
Comment (axic) [inline test/unittests/cxx20_bit_test.cpp:50]: @chfast how about this version? Got it from `leb128_tests`.

**Comment by chfast** [comment]
> > 
> 
> At this point though the tests should be changed to a pair then.

Not sure what exactly do you mean, but you can create a new header file with test cases for `clz, ctz, popcnt`:
```cpp
struct BitCountTestCase64
{
uint64_t value;
uint32_t expected_clz;
...
};

constexpr BitCountTestCase64 test_cases_64[] = {{0,32,32,0}, ...}; 

```

File: test/unittests/cxx20_bit_test.cpp (hunk 1 above)
Comment (chfast) [reply test/unittests/cxx20_bit_test.cpp:50]: This is fine, but I'd put these in a header to be reused for `i32_popcnt` instruction tests.

File: test/unittests/cxx20_bit_test.cpp [hunk 2]
```diff
@@ -3,6 +3,7 @@
 // SPDX-License-Identifier: Apache-2.0
 
 #include "cxx20/bit.hpp"
+#include <test/unittests/bitcount_test_cases.hpp>
```
Comment (chfast) [inline test/unittests/cxx20_bit_test.cpp:6]:
```suggestion
#include "bitcount_test_cases.hpp"
```

File: test/unittests/execute_numeric_test.cpp [hunk 3]
```diff
@@ -6,6 +6,7 @@
 #include "instructions.hpp"
 #include "parser.hpp"
 #include <gtest/gtest.h>
+#include <test/unittests/bitcount_test_cases.hpp>
```
Comment (chfast) [inline test/unittests/execute_numeric_test.cpp:9]:
```suggestion
#include "bitcount_test_cases.hpp"
```

Comment (axic) [reply test/unittests/execute_numeric_test.cpp:9]: Why is `parser.hpp`/etc. included without a path btw?

File: test/unittests/cxx20_bit_test.cpp (hunk 2 above)
Comment (axic) [reply test/unittests/cxx20_bit_test.cpp:6]: Done.

**Comment by axic** [comment]
Any more test cases to be added?

File: test/unittests/cxx20_bit_test.cpp (hunk 1 above)
Comment (axic) [reply test/unittests/cxx20_bit_test.cpp]: Done.

File: test/unittests/cxx20_bit_test.cpp [hunk 4]
```diff
@@ -47,20 +48,12 @@ TEST(cxx20_bit, bit_cast_uint32_to_array)
...
 
 TEST(cxx20_bit, popcount64)
 {
-    EXPECT_EQ(popcount(uint64_t{0}), 0);
-    EXPECT_EQ(popcount(uint64_t{0xffffffffffffffff}), 64);
-    EXPECT_EQ(popcount(uint64_t{0xffffffff00000000}), 32);
-    EXPECT_EQ(popcount(uint64_t{0x00000000ffffffff}), 32);
-    EXPECT_EQ(popcount(uint64_t{0x0000ffffffff0000}), 32);
-    EXPECT_EQ(popcount(uint64_t{0x00ff00ff00ff00ff}), 32);
+    for (const auto& [input, expected] : test::popcount64_test_cases)
```
Comment (axic) [inline test/unittests/cxx20_bit_test.cpp:57]: Should we have `using namespace test;` at the top instead?

File: test/unittests/execute_numeric_test.cpp (hunk 3 above)
Comment (chfast) [reply test/unittests/execute_numeric_test.cpp]: They are not "public" so we just add "internal fizzy include" path to test targets. Which is `lib/fizzy`.

File: test/unittests/cxx20_bit_test.cpp (hunk 4 above)
Comment (chfast) [reply test/unittests/cxx20_bit_test.cpp:57]: Can be. Up to you.

**Comment by chfast** [comment]
> Any more test cases to be added?

Looks ok for now.

File: test/unittests/cxx20_bit_test.cpp (hunk 4 above)
Comment (axic) [reply test/unittests/cxx20_bit_test.cpp]: Did that then.

**Review by chfast — APPROVED** [review]

File: test/unittests/execute_numeric_test.cpp [hunk 5]
```diff
@@ -283,7 +284,8 @@ TEST(execute_numeric, i32_ctz)
 TEST(execute_numeric, i32_popcnt)
 {
     const auto i32_popcnt = create_unary_operation_executor(Instr::i32_popcnt);
-    EXPECT_THAT(i32_popcnt(0x7fff00), Result(7 + 8));
+    for (const auto& [input, expected] : popcount32_test_cases)
+        EXPECT_THAT(i32_popcnt(input), Result(expected));
```
Comment (gumb0) [inline test/unittests/execute_numeric_test.cpp:288]: To know what failed
```suggestion
        EXPECT_THAT(i32_popcnt(input), Result(expected)) << input;
```

File: test/unittests/cxx20_bit_test.cpp [hunk 6]
```diff
@@ -47,20 +48,12 @@ TEST(cxx20_bit, bit_cast_uint32_to_array)
...
 TEST(cxx20_bit, popcount32)
 {
-    EXPECT_EQ(popcount(uint32_t{0}), 0);
-    EXPECT_EQ(popcount(uint32_t{0xffffffff}), 32);
-    EXPECT_EQ(popcount(uint32_t{0xffff0000}), 16);
-    EXPECT_EQ(popcount(uint32_t{0x0000ffff}), 16);
-    EXPECT_EQ(popcount(uint32_t{0x00ffff00}), 16);
-    EXPECT_EQ(popcount(uint32_t{0x00ff00ff}), 16);
+    for (const auto& [input, expected] : popcount32_test_cases)
+        EXPECT_EQ(popcount(input), expected);
```
Comment (gumb0) [inline test/unittests/cxx20_bit_test.cpp:52]:
```suggestion
        EXPECT_EQ(popcount(input), expected) << input;
```

Comment (axic) [reply test/unittests/cxx20_bit_test.cpp:52]: Added.

Comment (gumb0) [reply test/unittests/cxx20_bit_test.cpp:52]: In this file still not added.

Comment (axic) [reply test/unittests/cxx20_bit_test.cpp:52]: Done.

**Review by gumb0 — APPROVED** [review]

**Pull request merged**